pkg/dependency/parser/python/uv/parser.go

```diff
@@ -0,0 +1,164 @@
+package uv
+
+import (
+	"github.com/BurntSushi/toml"
+	"github.com/samber/lo"
+	"golang.org/x/xerrors"
+
+	"github.com/aquasecurity/trivy/pkg/dependency"
+	ftypes "github.com/aquasecurity/trivy/pkg/fanal/types"
+	xio "github.com/aquasecurity/trivy/pkg/x/io"
+)
+
+type Lock struct {
+	Packages []Package `toml:"package"`
+}
+
+func (l Lock) packages() map[string]Package {
+	return lo.SliceToMap(l.Packages, func(pkg Package) (string, Package) {
+		return pkg.Name, pkg
+	})
+}
+
+func (l Lock) directDeps() map[string]struct{} {
+	deps := make(map[string]struct{})
+	root, exists := l.root()
+	if !exists {
+		return deps
+	}
+	for _, dep := range root.Dependencies {
+		deps[dep.Name] = struct{}{}
+	}
+	return deps
+}
+
+func (l Lock) devDeps() map[string]struct{} {
+	devDeps := make(map[string]struct{})
+
+	root, ok := l.root()
+	if !ok {
+		return devDeps
+	}
+
+	packages := l.packages()
+	visited := make(map[string]struct{})
+
+	var walkDeps func(Package)
+	walkDeps = func(pkg Package) {
+		if _, ok := visited[pkg.Name]; ok {
+			return
+		}
+		visited[pkg.Name] = struct{}{}
+		for _, dep := range pkg.Dependencies {
+			depPkg, exists := packages[dep.Name]
+			if !exists {
+				continue
+			}
+			walkDeps(depPkg)
+		}
+	}
+
+	walkDeps(root)
+
+	for _, pkg := range packages {
+		if _, exists := visited[pkg.Name]; !exists {
+			devDeps[pkg.Name] = struct{}{}
+		}
+	}
```

Comment:
I don't want to do 2 cycles
what do you think about this?
```suggestion
	var walkDeps func(string)
	walkDeps = func(pkgName string) {
		// Skip already visited packages
		if _, visitedFound := visited[pkgName]; visitedFound {
			return
		}
		// Skip dependency, if package with this name doesn't exist
		pkg, packageFound := packages[pkgName]
		if !packageFound {
			return
		}
		visited[pkg.Name] = struct{}{}
		for _, dep := range pkg.Dependencies {
			walkDeps(dep.Name)
		}
	}

	walkDeps(root.Name)
```

Reply:
UURC usually don't use nested functions, using them for similar cases (e.g. https://github.com/aquasecurity/trivy/blob/fd9ed3a330bc66e229bcbdc262dc296a3bf01f54/pkg/dependency/parser/nodejs/pnpm/parse.go#L230).
Can we move this function out?

Reply:
The suggested code does not use `devDeps`

Reply:
oh, I'm a little dumb, don't pay attention to the code provided, it really doesn't work

Reply:
Sorry, I forgot to write 2nd part my suggest 😄 
we can save non-dev dependencies and check `!ok` here:
https://github.com/aquasecurity/trivy/blob/ea1ac4ae991414f3cda73d55ac230831ff8c2507/pkg/dependency/parser/python/uv/parser.go#L125-L127

Reply:
for this `devDeps` needs to be filled in. I start a recursive traversal from the root project and its direct dependencies to get a list of all dependencies that don't belong to any group. All dependencies that are not visited during the traversal are dev.

Reply:
> All dependencies that are not visited during the traversal are dev.

You are absolutely right. Essentially, I'm suggesting that you reverse your logic:

```Go
func (l Lock) prodDeps() map[string]struct{} { // you can select another name
	root, ok := l.root()
	if !ok {
		return nil
	}

	packages := l.packages()
	visited := make(map[string]struct{})

	var walkDeps func(string)
	walkDeps = func(pkgName string) {
		// Skip already visited packages
		if _, visitedFound := visited[pkgName]; visitedFound {
			return
		}
		// Skip dependency, if package with this name doesn't exist
		pkg, packageFound := packages[pkgName]
		if !packageFound {
			return
		}
		visited[pkg.Name] = struct{}{}
		for _, dep := range pkg.Dependencies {
			walkDeps(dep.Name)
		}
	}

	walkDeps(root.Name)

	return visited
}

func (p *Parser) Parse(r xio.ReadSeekerAt) ([]ftypes.Package, []ftypes.Dependency, error) {
	var lock Lock
	if _, err := toml.NewDecoder(r).Decode(&lock); err != nil {
		return nil, nil, xerrors.Errorf("failed to decode uv lock file: %w", err)
	}

	packages := lock.packages()
	directDeps := lock.directDeps()
	prodDeps := lock.prodDeps()

	var (
		pkgs []ftypes.Package
		deps []ftypes.Dependency
	)

	for _, pkg := range lock.Packages {
		if _, ok := prodDeps[pkg.Name]; !ok { // See this line (`ok` changed to `!ok`)
			continue
		}
...
```

Reply:
That's a good idea. Done https://github.com/aquasecurity/trivy/pull/8080/commits/11c532e04813ed000346207d0e41675a5d317204